### DISCUSSION TITLE
Secrets Exposed When Passed Through Expressions

### DISCUSSION THREAD

**Discussion by ethanxr** [community/community#38323] [Actions] [answered]
If a Github secret is used in a workflow but passed through an expression such as `fromJSON`, the secret, or a portion of it, can be exposed in logs and debugging information.

For example, a workflow with the following
```
      env: 
        KEY_ID: ${{ fromJSON(secrets.MY_SECRET_JSON).key }}
```
This will log the secret's `key` value in plain and clear text instead of the redacted stars.

I understand that it may not be possible to redact secret data if it is not securely handled by scripts, but the redaction functionality is part of the Github Actions framework, so I would expect that the provenance of secrets are considered when using other parts of the Actions framework, such as expressions.

Is this expected behavior, or a security bug?

**Comment by airtower-luna** [comment, marked as answer]
The [documentation on secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets) explicitly warns about this:

> To help ensure that GitHub redacts your secret in logs, avoid using structured data as the values of secrets. For example, avoid creating secrets that contain JSON or encoded Git blobs.

So I'd call it a known limitation.
